pkg/conference/publisher/publisher.go

```diff
@@ -0,0 +1,116 @@
+package publisher
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/pion/rtp"
+	"github.com/sirupsen/logrus"
+)
+
+var ErrSubscriptionExists = errors.New("subscription already exists")
+
+type Subscription interface {
```

Comment:
this can be private like:

```
type subscription interface
```

Reply:
Will it work? I.e. the `Publisher.AddSubscription()` function and `Publisher.RemoveSubscription()` are both public functions that accept `subscription Subscription`.

Reply:
Yes, because a struct doesn’t have to implement an interface in Go. Go use structural typing (duck typing). This follows the Interface segregation principle. In Go interfaces should be small as possible and they not describe what you offer they describe what you need.

Reply:
Right, yet I still don't think making an interface private would work. `Subscription` interface is used in a publicly-exposed function which means that the types that it uses must be public as well, otherwise we're using unexported/private types in the exported/public functions.

I quickly tried to change the type to private and indeed it stopped compiling.

Reply:
I changed this in this way, and it compiling.

```go
type subscription interface {
	// WriteRTP **must not** block (wait on I/O).
	WriteRTP(packet rtp.Packet) error
}
```

Reply:
Ah, right, it compiles as well, I confused it with the `gopls` warning that `subscription` would shadow a variable with the same name.

Interesting! But do the callers to `AddSubscription`, `RemoveSubscription` know about the private type? I.e. if I were to use the `publisher` package on its own, would I as a caller know the requirements of the interface despite it being private?

(I'm now wondering what's the difference between public and private interfaces then, is it only different in a way that public interfaces could be created in an external code that uses the package? i.e. in case we were **to return** the `subscription`, we would need to make it public)

Reply:
I think we're having the same conversation in two places. Please see my comment on 'accept interfaces return structs'. I would mark this thread as solved. In general I'm fine with the PR

Reply:
To your example:
I think a caller with a hash-map of subscriptions should knows (define by him self) the structure about what the he save. 
The publisher only have to know thats a subscriptions has a method: `WriteRTP(packet rtp.Packet) error` thats it.
The Caller should not ask the Publisher what he knows about subscriptions?

Maybe the name of this subscription interface is miss leading. In idiomatic Go we would name the interface like:

```Go
type rtpWriter interface {
	// WriteRTP **must not** block (wait on I/O).
	WriteRTP(packet rtp.Packet) error
}
```

Reply:
Maybe another point to this view. (Even if I'm starting a fundamental discussion here 👀 )

Patterns like Factory do not makes sense in Go.

Reply:
> The second point was, how we deal with signal channels from parent go routines.
> A child should tell parent routine about close than reaction on parent routine closing.

Do you mean that the parent should never inform the child that they're done, signaling the child to stop?

> I think a caller with a hash-map of subscriptions should knows (define by him self) the structure about what the he save.

That is true, yet the caller may (for whatever reason) store the list of all available subscriptions in a collection, let's say in a map or in an array. The map or an array can be a collection of heterogenous types, all of which only have one thing in common: they are all subscriptions. This way, the caller may store them `map[string]publisher.Subscription`, otherwise there will be no other option apart from `map[string]interface{}` to store heterogeneous types in a single collection.

We as publishers don't care about the exact type or the implementation details of the subscriptions, we're ok accepting any type as long as it satisfies the contract (the interface), the actual subscription could be anything from a WebRTC track to just plain RTP sink or any other stream that accepts rtp packets 🙂 

> Patterns like Factory do not makes sense in Go.

Sorry, I did not get what you meant by it (in particular, what it related to 🙂).